design/0001-frr.md

```diff
@@ -0,0 +1,411 @@
+provides detailed logging features to aid debugging.
+
+The current native BGP implementation of MetalLB has limitations when coexisting
+with other BGP agents on the node. If these BGP agents are attempting to peer
+with external routers, this may invoke BGP’s conflict resolution algorithm. This
+is detailed [here](https://metallb.universe.tf/configuration/calico/) for the
+Calico case and [here](https://metallb.universe.tf/configuration/kube-router/)
+for the Kube-Router case. This enhancement may be a step towards resolving these
+issues by potentially allowing MetalLB to reuse a BGP agent already in use on
+the node. This would enable the local BGP agent (FRR) to resolve any conflicts
+locally before advertising to a remote BGP peer and, therefore, becoming the
+single point-of-contact with that peer.
+
+### Goals
+
+* Provide a configuration option to allow use of FRR as a BGP agent in a MetalLB
+  deployment at cluster installation time.
+* Ensure feature parity with current BGP implementation.
+* Reinstate thorough BGP testing infrastructure.
+* Add additional documentation to indicate how to troubleshoot and debug any
+  issues with FRR.
+* This will not expose direct configuration of FRR to the MetalLB user.
+
+### Non-Goals
+
+* Coexistence of two types of BGP agents (e.g. MetalLB native BGP agent AND FRR
+  BGP agent) is not a goal of this work.
+* The purpose of this is not to provide a method to integrate arbitrary BGP
+  stacks into MetalLB. However, this work will develop a well-defined
+  interface that could be used to integrate other BGP stacks and will be a
+  template to do so.
+* This will not support runtime changing of BGP agent type.
+* This will not extend the functionality of the MetalLB BGP implementation in
+  in order to take advantage of the additional cababilities of FRR. This will
+  be a goal of future enhancements.
+* This will not explicitly support allowing another actor to interact with FRR.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a developer, I want to be able to run a comprehensive set of BGP system
+tests, preferably based on kubernetes e2e tests, in order to ensure the MetalLB
+BGP implementation works as expected.
+
+#### Story 2
+
+As a cluster administrator, I want to be able to specify a configuration option
+when deploying MetalLB in order to select the MetalLB BGP agent to be used by
+MetalLB
+
+#### Story 3
+
+As a developer, I want to be able to configure FRR from MetalLB using FRR Go
+bindings in order to improve performance, reliability and debugging
+capabilities.
+
+### Risks and Mitigations
+
+FRR does not provide documented northbound Go bindings to allow configuration of
+BGP. There is an experimental gRPC interface. This interface may need to be
+productionized through the FRR community. When this has been satisfactorily
+achieved, we can start [Story 3](#story-3). Until that time, in order to
+mitigate this risk, we can wrap the FRR ‘vtysh’ command line interface in Go
```

Comment:
Does wrapping `vtysh` entail automating an interactive CLI using Go code? This sounds fragile to me. I'd opt for updating a config file if FRR supports live config reloading until we have e.g. gRPC support.

Reply:
I wasn't considering wrapping the interactive `vtysh` shell but rather send commands using vtysh which should be less fragile. i.e. `vtysh -c "command"`.

Using a config file is an option. There are some limitations with that that would be preferable to address (https://github.com/FRRouting/frr/issues/2128)

Reply:
Shall I update the document to state the preferred method (e.g. `vtysh -c cmd`)?

Reply:
LGTM. @johananl ?